api/oc_push.c

```diff
@@ -0,0 +1,2625 @@
+
+  /* look up target receivers of target Push Receiver Resource */
+  recvs_instance = (oc_recvs_t *)oc_list_head(recvs_list);
+  while (recvs_instance) {
+    if (recvs_instance->resource == request->resource) {
+      OC_PUSH_DBG("receivers obj array instance of push receiver resource "
+                  "(\"%s\") is found!",
+                  oc_string(request->resource->uri));
+
+      if (uri_param_len != -1) {
+        recv_obj =
+          _find_recv_obj_by_uri(recvs_instance, uri_param, uri_param_len);
+        if (recv_obj) {
+          /* remove receiver obj from array */
+          oc_list_remove(recvs_instance->receivers, recv_obj);
+          OC_PUSH_DBG("receiver obj is removed from array");
+
+          /* delete associated resource... */
+          _purge_pushd_rsc(&recv_obj->receiveruri,
+                           recvs_instance->resource->device);
+          OC_PUSH_DBG(
+            "app resource corresponding to the receiver obj is removed");
+
+          /* free memory */
+          oc_free_string(&recv_obj->receiveruri);
+          oc_free_string_array(&recv_obj->rts);
+          oc_memb_free(&recv_instance_memb, recv_obj);
+        } else {
+          /* if the given `receiveruri` parameter is not in existing receivers
+           * array, add new receiver object to the receivers array */
+#ifdef OC_PUSHDEBUG
+          //					oc_string_t uri;
+          //					oc_new_string(&uri, uri_param, uri_param_len);
+          OC_PUSH_DBG(
+            "can't find receiver object which has uri(\"%.*s\"), ignore it...",
+            uri_param_len, uri_param);
+//					oc_free_string(&uri);
+#endif
+          result = OC_STATUS_NOT_FOUND;
+        }
+      } else {
+        /* if `receiveruri` param is not provided..
+         * remove whole existing `receivers` object array */
+        _purge_recv_obj_list(recvs_instance);
+      }
+
+      break;
+    }
+
+    recvs_instance = recvs_instance->next;
+  }
+
+  oc_send_response(request, result);
+
+  return;
+}
+
+/*
+ * @brief	initiate Push Receiver Resource
+ *
+ * @details
+ * for Target Server \n
+ * - Push Receiver ("oic.r.pushreceiver") \n
+ *
+ * @param device_index  device index
+ */
+void
+oc_create_pushreceiver_resource(size_t device_index)
+{
+  /* create Push Receiver Resource */
+  oc_resource_t *push_recv =
+    oc_new_resource("Push Receiver", PUSHRECVS_RSC_PATH, 1, device_index);
+
+  if (push_recv) {
+    oc_resource_bind_resource_type(push_recv, "oic.r.pushreceiver");
+    oc_resource_bind_resource_interface(push_recv, OC_IF_RW | OC_IF_BASELINE);
+    oc_resource_set_default_interface(push_recv, OC_IF_RW);
+    oc_resource_set_discoverable(push_recv, true);
+
+    oc_resource_set_request_handler(push_recv, OC_GET, get_pushrecv, NULL);
+    oc_resource_set_request_handler(push_recv, OC_POST, post_pushrecv, NULL);
+    oc_resource_set_request_handler(push_recv, OC_DELETE, delete_pushrecv,
+                                    NULL);
+
+    oc_add_resource(push_recv);
```

Comment:
We can save some instructions by moving this to line 2218:

```
    if (recvs_instance) {
      oc_add_resource(push_recv);
      recvs_instance->resource = push_recv;
      ...
    } else {
    ...
    }
```

The `oc_delete_resource` call in the else branch won't need to remove the resource from the app_resources list, which saves some instructions. :-)

Reply:
agreed, fixed

Reply:
ok, done